stun/stunhmac-openssl.c

```diff
@@ -0,0 +1,88 @@
+#include <assert.h>
+
+#include <openssl/md5.h>
+#include <openssl/hmac.h>
+
+#include "rand.h"
+#include "stunhmac.h"
+
+void stun_sha1 (const uint8_t *msg, size_t len, size_t msg_len,
+    uint8_t *sha, const void *key, size_t keylen, int padding) {
+  uint16_t fakelen = htons(msg_len);
+  uint8_t pad_char[64] = {0};
+  HMAC_CTX handle;
+
+#ifdef NDEBUG
+#define TRY(x) x;
+#else
+  int ret;
+#define TRY(x)                                  \
+  ret = x;                                      \
+  assert (ret >= 0);
+#endif
+
+  assert (len >= 44u);
+
+  HMAC_CTX_init(&handle);
+  TRY(HMAC_Init(&handle, key, keylen, EVP_sha1()));
+
+  TRY(HMAC_Update(&handle, msg, 2));
+  TRY(HMAC_Update(&handle, (unsigned char*)&fakelen, 2));
+  TRY(HMAC_Update(&handle, msg + 4, len - 28));
+
+  /* RFC 3489 specifies that the message's size should be 64 bytes,
+     and \x00 padding should be done */
+  if (padding && ((len - 24) % 64) > 0) {
+    uint16_t pad_size = 64 - ((len - 24) % 64);
+
+    TRY(HMAC_Update(&handle, pad_char, pad_size));
+  }
+
+  unsigned int shalen;
+  HMAC_Final(&handle, sha, &shalen);
+  assert(shalen == 20);
+
+  HMAC_CTX_cleanup(&handle);
+  HMAC_cleanup(&handle);
+}
+
+static const uint8_t *priv_trim_var(const uint8_t *var, size_t *var_len)
+{
+  const uint8_t *ptr = var;
+
+  while (*ptr == '"') {
+    ptr++;
+    (*var_len)--;
+  }
+  while(ptr[*var_len-1] == '"' ||
+      ptr[*var_len-1] == 0) {
```

Comment:
Similarly, we should break here if we ever get down to `*var_len == 0`.  I don't know that this condition is reachable unless we're called with `*var_len == 0` initially, but it def makes me more comfortable